src/protocol/oprf/sharded_shuffle/shard_picker.rs

```diff
@@ -0,0 +1,108 @@
+use rand::{
+    distributions::{WeightedError, WeightedIndex},
+    prelude::Distribution,
+    Rng,
+};
+
+#[derive(Debug, Clone, Copy)]
+struct ShardInfo {
+    empty_slots: usize,
+}
+
+struct ShardPicker {
+    total_available_slots: usize,
+    shards: Box<[ShardInfo]>,
+    dist: WeightedIndex<usize>,
+}
+
+impl ShardPicker {
+    pub fn new(num_shards: usize, shard_size: usize) -> Result<Self, WeightedError> {
```

Comment:
Could you please explain to me the benefits of this optimization?

1. All operation on Vec/slices like indexed access are perfomed on usizes anyway. So, even if I  accept num_shards as a u16 it will be cast into usize on each iteration anyway.
2. If we expect the number of shards to not be greater than 64K, then is the usize->u32 optimization for shard_size really worth? It is going to save no more than 64K * (8-4) = 256Kb of memory? Is it because of something related to fitting this all into CPU caches provide the code to be in a hot path?
3. Also, provided that shard picking is going to be in a hot path, isn't it better to let CPU operate on 64bit values (equal to register sizes), rather than let it deal with 64->32 bit conversions?

Basically I'd like to learn you line of thought here.

Reply:
I am not suggesting to do this for the purpose of making code more efficient, but rather use more appropriate types to represent this information. number of shards and shard size should not be platform-specific and 8 bytes is too much for either of those arguments imo. 

It is possible and probably ok to use `usize` internally, but rejecting invalid values at compile type by requesting appropriately-sized types for function arguments saves a lot of headache in the future.

> Also, what does "infallible" mean here? Is it some Rust-specific lingo?

Infallible functions don't fail. In this case, if arguments provided to it are guaranteed to be positive non-zero integers, then it is always possible to make this struct. `NonZeroUXX` type can help here.